File: src/main/java/org/ybiquitous/messages/MessageKey.java
package org.ybiquitous.messages;

import java.util.Locale;

public final class MessageKey {

    public static MessageKey of(String key) {
        return of(key, DEFAULT_RESOURCE_NAME);
    }

    public static MessageKey of(String key, String resourceName) {
        return of(key, DefaultMessageBuilder.create(resourceName));
    }

    public static MessageKey of(String key, MessageBuilder builder) {
        return new MessageKey(key, builder);
    }

    private static final String DEFAULT_RESOURCE_NAME = "messages";

    private final String _key;

    private transient final MessageBuilder _builder;

    private MessageKey(String key, MessageBuilder builder) {
        this._key = Utils.notNull(key, "key");
        this._builder = Utils.notNull(builder, "builder");
    }

    public String get(Object... args) {
        return get(MessageLocaleHolder.get(), args);
    }

    public String get(Locale locale, Object... args) {
        return this._builder.build(locale, this._key, args);
    }

    @Override
    public String toString() {
        return this._key;
    }

    @Override
    public int hashCode() {
        return this._key.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        MessageKey that = (MessageKey) obj;
        return this._key.equals(that._key);
    }
}

File: src/main/scala/org/ybiquitous/messages/scala/MessageKey.scala
package org.ybiquitous.messages.scala

import java.util.Locale

import org.ybiquitous.messages.MessageBuilder
import org.ybiquitous.messages.{ MessageKey => MsgKey }

final class MessageKey private(val delegate: MsgKey) {
  def get(args: Any*) = delegate.get(args map unwrapArg: _*)
  def get(locale: Locale, args: Any*) = delegate.get(locale, args map unwrapArg: _*)
  override def toString = delegate.toString
  override def hashCode = delegate.hashCode
  override def equals(obj: Any) = obj match {
    case that: MessageKey => this.delegate == that.delegate
    case _ => false
  }
  private def unwrapArg(arg: Any): AnyRef = arg.asInstanceOf[AnyRef]
}

object MessageKey {
  def apply(
    key: String,
    resourceName: Option[String] = None,
    builder: Option[MessageBuilder] = None
  ): MessageKey = {
    if (resourceName.isEmpty && builder.isEmpty) {
      return new MessageKey(MsgKey.of(key))
    }
    if (!resourceName.isEmpty && builder.isEmpty) {
      return new MessageKey(MsgKey.of(key, resourceName.get))
    }
    if (resourceName.isEmpty && !builder.isEmpty) {
      return new MessageKey(MsgKey.of(key, builder.get))
    }
    if (!resourceName.isEmpty && !builder.isEmpty) {
      return new MessageKey(MsgKey.of(key, resourceName.get))
    }
    throw new AssertionError("never reach: " + List(key, resourceName, builder) )
  }
}
